### DISCUSSION TITLE
Some questions about external_petsc_problem 2.0?

### DISCUSSION THREAD

**Discussion by Wang-Yihu** [idaholab/moose#29369] [Q&A General] [answered]
### Check these boxes if you have followed the posting rules.

- [X] Q&A General is the most appropriate section for my question
- [X] I have consulted the posting Guidelines on the Discussions front page
- [X] I have searched the Discussions forum and my question has not been asked before
- [X] I have searched the MOOSE website and the documentation does not answer my question
- [X] I have formatted my post following the posting guidelines (screenshots as a last resort, triple back quotes around pasted text)

### Question

Still the question about this part `\modules\external_petsc_solver\test\tests\external_petsc_problem\moose_as_parent.i`
Q2: What makes the Picard iteration? What's the meaning of the Picard residual in the `moose_as_parent.i` and `petsc_transient_as_sub.i`?
When I run `moose_as_parent.i`, it will execute following these steps
```
Time Step 0, time = 0

Time Step 1, time = 0.2
Beginning fixed point iteration 0
sub_app0: ...
Main app solve::...
Picard:...
Beginning fixed point iteration 1
sub_app0: ...
Main app solve::...
Picard:...
Beginning fixed point iteration 2
sub_app0: ...
Main app solve::...
Picard:...

Time Step 2, time = 0.4
Beginning fixed point iteration 0
sub_app0: ...
Main app solve::...
Picard:...
Beginning fixed point iteration 1
sub_app0: ...
Main app solve::...
Picard:...
Beginning fixed point iteration 2
sub_app0: ...
Main app solve::...
Picard:...

Time Step 3, time = 0.6
Beginning fixed point iteration 0
sub_app0: ...
Main app solve::...
Picard:...
Beginning fixed point iteration 1
sub_app0: ...
Main app solve::...
Picard:...
Beginning fixed point iteration 2
sub_app0: ...
Main app solve::...
Picard:...
....................................
```

Certainly, I find the external solve must be conducted by `void ExternalPETScProblem::externalSolve()`.
But what conducts the Picard iteration? Is it from` void ExternalPETScProblem::advanceState()` ? Also, it seems that there is no data transfer from MOOSE to external solver in this example. In other words, the condition of the start for each time step of the external solver only following its way and does not have any relation with MOOSE main solver, so why does it need Picard iteration? What's the meaning of the residual of the Picard iteration?

**Comment by GiudGiud** [comment]
Hello

The fixed point iterations are performed by the FixedPointSolve class.
You can see that when executing the moose_as_parent.i input file, you use a Transient executioner
```

[Executioner]
  type = Transient
  num_steps = 10
  dt = 0.2

  solve_type = 'PJFNK'

  fixed_point_max_its = 10
  fixed_point_rel_tol = 1e-8
  fixed_point_abs_tol = 1e-9
  nl_rel_tol = 1e-6
  nl_abs_tol = 1e-12

  petsc_options_iname = '-pc_type -pc_hypre_type'
  petsc_options_value = 'hypre boomeramg'
[]
```

because of the fixed_point parameters that were passed, a FixedPointSolve class is created to wrap both the MultiApps and the FEProblemSolve (the default solve for this input file) in a fixed point iteration

**Reply by GiudGiud** [reply]
>   Also, it seems that there is no data transfer from MOOSE to external solver in this example. In other words, the condition of the start for each time step of the external solver only following its way and does not have any relation with MOOSE main solver, so why does it need Picard iteration?
You're right there is no Transfers to the child app in this input file
Thus it does not need a Picard iteration.

> What's the meaning of the residual of the Picard iteration?

It's the nonlinear residual of the parent app. Here it decreases every time the solver is called. The first time it is called, it only converged to the requested tolerance `  nl_rel_tol = 1e-6` then the next time it's uses the updated v field and converges firther, on the 3rd solve it is already tight and the `v` field does not change again, so  that's why the residual is very small

**Reply by Wang-Yihu** [reply]
Is this your meaning for every time step? 
Sub is executed in the beginning.
First in parent, it uses v = 0 or v from the previous time step .
Second in parent, it uses v from the next time step.
Third in parent, the v does not change. 
But if this is true, the first, second, and third cut a time step into 3 parts? Actually they do not represent the full time step?
![image](https://github.com/user-attachments/assets/d121d140-3d0e-46ca-8f22-03871ab80c82)

**Reply by GiudGiud** [reply]
> Sub is executed in the beginning.

TransientMultiApp is executed on TIMESTEP_BEGIN by default.
But that's not what you show here. It should be:
- sub executes
- data transferred to main
- main executes
- sub executes
- data transferred to main
- main executes with a tigther convergence

you can just look at the output log on the console, it shows that
```

Time Step 1, time = 0.2, dt = 0.2
Initial fixed point residual norm: 3.316625e+00

Beginning fixed point iteration 0

sub_app0: 
sub_app0: Time Step 1, time = 0.2, dt = 0.2
sub_app0: syncSolutions to external petsc App 
sub_app0: PETSc External Solve!
0 TS dt 0.2 time 0.
    0 SNES Function norm 7.215811749992e+02
      0 KSP Residual norm 1.812600607724e+00
      1 KSP Residual norm 3.184000953610e-01
      2 KSP Residual norm 1.109782762250e-01
      3 KSP Residual norm 4.774287620076e-02
      4 KSP Residual norm 1.066119625877e-02
      5 KSP Residual norm 1.905636433419e-03
      6 KSP Residual norm 2.071925087269e-04
      7 KSP Residual norm 4.123842437916e-05
      8 KSP Residual norm 9.426200921053e-06
    1 SNES Function norm 3.009696552534e-03
      0 KSP Residual norm 9.426200920969e-06
      1 KSP Residual norm 2.660932951319e-06
      2 KSP Residual norm 6.542299529123e-07
      3 KSP Residual norm 8.976376835224e-08
      4 KSP Residual norm 1.929475978757e-08
      5 KSP Residual norm 8.631984111349e-09
      6 KSP Residual norm 2.470485970064e-09
      7 KSP Residual norm 4.240939651349e-10
      8 KSP Residual norm 6.723165717722e-11
    2 SNES Function norm 2.268243483623e-08
1 TS dt 0.2 time 0.2
sub_app0: syncSolutions from external petsc App
sub_app0:  Solve Converged!
Fixed point residual norm after TIMESTEP_BEGIN MultiApps: 1.951095e+01

Main app solve:
 0 Nonlinear |R| = 1.969494e+01
      0 Linear |R| = 1.969494e+01
      1 Linear |R| = 1.574952e+00
      2 Linear |R| = 2.764162e-02
      3 Linear |R| = 3.807315e-04
      4 Linear |R| = 5.703901e-06
 1 Nonlinear |R| = 1.007532e-05
 Solve Converged!

 0 Picard |R| = 3.316625e+00
 1 Picard |R| = 1.951095e+01
```

> But if this is true, the first, second, and third cut a time step into 3 parts? Actually they do not represent the full time step?

No the time step is not cut. It's just the same time step taken several times, with just a tighter convergence criterion every time. It's because it relies on a loose relative convergence, so every execution is relative to the residual achieved on the previous solve

**Reply by Wang-Yihu** [reply, marked as answer]
So is this your meaning
![image](https://github.com/user-attachments/assets/1941947d-dadf-4146-b244-49510ac6cdec)
for the same time step different fixed point iterations: 
sub - the same; 
main - the value in the following fixed point iterations step inherits the value in the previous fixed point iterations step. In the following fixed point iterations step, the iteration may continue for another relative tolerance or absolute tolerance with no time promotion. 
And the meaning of the picard |R|: just non-liner residual for the main...

**Reply by GiudGiud** [reply]
Yes the figure is accurate
see similar things here
https://mooseframework.inl.gov/getting_started/examples_and_tutorials/tutorial02_multiapps/presentation/index.html#/4/6

> And the meaning of the picard |R|: just non-liner residual for the main...

yes. We are upgrading this soon using a new Convergence system which will allow us to define arbitrary convergence criteria
